### PR TITLE
gh-122208: Don't delivery PyDict_EVENT_ADDED until it can't fail

### PR DESCRIPTION
**Pull request by DinoV**
Currently `PyDict_EVENT_ADDED` can be delivered when the insert is going to fail due to an OOM. This means the listener has no capability of knowing the true state of the dictionary if it fails.

This modifies the events to be delivered before the change has happened but only after any underlying necessary allocations have occurred.

* Issue: gh-122208

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -1745,16 +1750,11 @@ insertdict(PyInterpreterState *interp, PyDictObject *mp,
     MAINTAIN_TRACKING(mp, key, value);
 
     if (ix == DKIX_EMPTY) {
-        assert(!_PyDict_HasSplitTable(mp));
```
Comment (carljm) [inline Objects/dictobject.c:1748]: Was it intentional to remove this assertion? Doesn't seem related to the dict watcher stuff.

**Review by carljm — APPROVED** [review]
Looks good, thanks for the fix! Sorry about the debugging session that I presume must have led to this find :)

File: Objects/dictobject.c (hunk 1 above)
Comment (DinoV) [reply Objects/dictobject.c:1748]: No, thanks for catching that, I'll remove it.

Luckily there was no debugging involved! Just trying to re-factor some existing code to use dict watchers and I was worried about what would happen when the updates would fail (as the existing code handled it). So this will actually make that somewhat cleaner!
